core/raftlease/fsm.go

```diff
@@ -0,0 +1,411 @@
+	start time.Time
+
+	// duration is the duration for which the lease is valid,
+	// from the start time.
+	duration time.Duration
+}
+
+// Apply is part of raft.FSM.
+func (f *FSM) Apply(log *raft.Log) interface{} {
+	var command Command
+	err := yaml.Unmarshal(log.Data, &command)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := command.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+	switch command.Operation {
+	case OperationClaim:
+		return f.claim(command.LeaseKey(), command.Holder, command.Duration)
+	case OperationExtend:
+		return f.extend(command.LeaseKey(), command.Holder, command.Duration)
+	case OperationExpire:
+		return f.expire(command.LeaseKey())
+	case OperationSetTime:
+		return f.setTime(command.OldTime, command.NewTime)
+	default:
+		return errors.NotValidf("operation %q", command.Operation)
+	}
+}
+
+// Snapshot is part of raft.FSM.
+func (f *FSM) Snapshot() (raft.FSMSnapshot, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	entries := make(map[SnapshotKey]SnapshotEntry, len(f.entries))
+	for key, entry := range f.entries {
+		entries[SnapshotKey{
+			Namespace: key.Namespace,
+			ModelUUID: key.ModelUUID,
+			Lease:     key.Lease,
+		}] = SnapshotEntry{
+			Holder:   entry.holder,
+			Start:    entry.start,
+			Duration: entry.duration,
+		}
+	}
+	return &Snapshot{
+		Version:    SnapshotVersion,
+		Entries:    entries,
+		GlobalTime: f.globalTime,
+	}, nil
+}
+
+// Restore is part of raft.FSM.
+func (f *FSM) Restore(reader io.ReadCloser) error {
+	defer reader.Close()
+	var snapshot Snapshot
+	decoder := yaml.NewDecoder(reader)
+	if err := decoder.Decode(&snapshot); err != nil {
+		return errors.Trace(err)
+	}
+	if snapshot.Version != SnapshotVersion {
+		return errors.NotValidf("snapshot version %d", snapshot.Version)
+	}
+	if snapshot.Entries == nil {
+		return errors.NotValidf("nil entries")
+	}
+
+	newEntries := make(map[lease.Key]*entry, len(snapshot.Entries))
+	for key, ssEntry := range snapshot.Entries {
+		newEntries[lease.Key{
+			Namespace: key.Namespace,
+			ModelUUID: key.ModelUUID,
+			Lease:     key.Lease,
+		}] = &entry{
+			holder:   ssEntry.Holder,
+			start:    ssEntry.Start,
+			duration: ssEntry.Duration,
+		}
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	f.globalTime = snapshot.GlobalTime
+	f.entries = newEntries
+
+	return nil
+}
+
+// Snapshot defines the format of the FSM snapshot.
+type Snapshot struct {
+	Version    int                           `yaml:"version"`
+	Entries    map[SnapshotKey]SnapshotEntry `yaml:"entries"`
+	GlobalTime time.Time                     `yaml:"global-time"`
+}
+
+// Persist is part of raft.FSMSnapshot.
+func (s *Snapshot) Persist(sink raft.SnapshotSink) (err error) {
+	defer func() {
+		if err != nil {
+			sink.Cancel()
+		}
+	}()
+
+	encoder := yaml.NewEncoder(sink)
+	if err := encoder.Encode(s); err != nil {
+		return errors.Trace(err)
+	}
+	if err := encoder.Close(); err != nil {
+		return errors.Trace(err)
+	}
+	return sink.Close()
```

Comment:
This method gets called by the raft infrastructure - the docs say that the FSM code should call sink.Close when it's finished with it. We definitely have the same kind of pattern in our code where we need it - otherwise no-one would ever take a WriteCloser, right? :)